src/Course/StateT.hs

```diff
@@ -267,32 +268,38 @@ instance Functor f => Functor (OptionalT f) where
 instance Monad f => Applicative (OptionalT f) where
   pure = OptionalT . pure . Full
 
-  -- (<*>) :: OptionalT (f (Optional (a -> b))) -> OptionalT (f (Optional a)) -> OptionalT (f (Optional b))
-  -- ffunc :: f (Optional (a -> b))
-  -- fa    :: f (Optional a)
-  -- func <*> :: f _ -> f b
-  OptionalT ffunc <*> OptionalT fa = OptionalT $ (<*>) ffunc <*> _
-
-  -- OptionalT ffunc <*> OptionalT fa = OptionalT $ do
-  --   maybeFunc <- ffunc
-  --   maybeA <- fa
-  --   case maybeFunc of
-  --     Empty -> Empty
-  --     Full func -> case maybeA of
-  --       Empty -> Empty
-  --       Full a -> Full $ func a
-
-
-  -- OptionalT ffunc <*> OptionalT fa = onFull t -> f (Optional a) Optional t
-  --   Applicative f => (t -> f (Optional a)) -> Optional t -> f (Optional a)
-  -- theirs: OptionalT $ (<*>) <$> ffunc <*> fa (but doesn't pass all tests?)
+  OptionalT ffunc <*> OptionalT fa = OptionalT $ do
+    maybeFunc <- ffunc
+    case maybeFunc of
+      Empty     -> pure Empty
+      Full func -> do
+        a <- fa
+        pure $ func <$> a
+    -- REVIEW:
+    -- Above, you suggested solutions like these over more point-free ones like theirs
+    -- (OptionalT f <*> OptionalT a = OptionalT (lift2 (<*>) f a)). Does this fall into the same
+    -- category? I'm still trying to build my intuition around when it makes sense to go point-free
+    -- and when it doesn't.
+    --
+    -- Also, their version fails tests 5-7
```

Comment:
This has to do with how `lift2` is defined for lists. In order for these tests to pass, it has to do what your code above does: first run the first argument, then the second, not both at once. I'm not sure what your implementation here is, so can't say more.

In general, however, it is kind of well known that lists are good as functors, but very weird as applicatives and monads. This is because there are multiple ways of implementing `<*>` and `=<<` for lists, and none of them make intuitive sense in all edge cases.

Reply:
Actually, scratch that. There is no way to implement `lift2` on lists in a way that will accommodate these tests.

```
lift2 (+) [1,2] [10, 20] == [11, 12, 21, 22]
lift2 ($) [(+1), (+2)] [10, 20] == [11, 12, 21, 22]
```

It can't know that it has to skip application for some elements, unless it knows the inner structure of those elements, so for `lift2 (<*>) [Empty, Empty] [Full 1, Full 2]` it will always apply both `Empty`ies to both `Full`s, returning a list of 4.

In general you can say that:

```
∀f, l1, l2 : length (lift2 f l1 l2) == length f1 * length f2
```

There is, of course, another way to implement `lift2` on lists:

```
lift2 (+) [1,2] [10,20] == [11, 22]
```

But that will also not work with some of the tests.

Reply:
Is it fair to say that one should (almost) never see lists as applicatives or monads in production code?

Reply:
No.
Even though lists are not well-behaved monads, it is nevertheless very convenient to use the `do` notation for them:

```
cartesianProduct :: [a] -> [b] -> [(a, b)]
cartesianProduct ax bx = do
    a <- ax
    b <- bx
    pure (a, b)
```

Of course, in this specific case I could use a list comprehension, but there are more complicated cases where list comprehension comes out really ugly, and monadic syntax looks very understandable. And people do this all the time. In fact, PureScript doesn't even have list (array?) comprehensions, and instead recommends using the monadic syntax.

Reply:
So then what's the right way to think about this problem?
* It's aesthetically more pleasing to use `lift2`, but doing so doesn't meet the necessary requirements, so we have to use a slightly uglier solution
* The specs are defined wrong in some way, it should be possible to use `lift2` to solve this problem
* Something else?

Reply:
Either of the first two ways can apply. The requirements are arbitrary, there is no real way to make a judgement between the two approaches.

Reply:
I'm having trouble making sense of the combination of the following statements:

> In general, however, it is kind of well known that lists are good as functors, but very weird as applicatives and monads. This is because there are multiple ways of implementing `<*>` and `=<<` for lists, and none of them make intuitive sense in all edge cases.

> Even though lists are not well-behaved monads, it is nevertheless very convenient to use the `do` notation for them...In fact, PureScript doesn't even have list (array?) comprehensions, and instead recommends using the monadic syntax.

The former seems to say "list monads/applicatives sometimes have surprising behavior, and thus you probably shouldn't use them if you can avoid it" and the former seems to say "list monads/applicatives are used all the time and some people think they're a better tool than some other tools."

Reply:
It's an approximation thing. They behave well enough in mainstream cases. It's only edge cases that are weird.

It's like Newton's gravity is not strictly speaking correct, so you should ideally use relativity, but it is _correct enough_ under normal circumstances that you can still get away with it most of the time.

Reply:
How do I know when I'm getting near the edges of the safe uses of list monads/applicatives? Or is that just from experience?

Reply:
It depends on what you mean by "safe".
Technically, everything always works exactly the way it was programmed to.
When I say "weird", I mean that behavior of the program does not correspond to your intuitive expectation.

From here, one can argue that "safe" means "corresponding to my intuition", and therefore the way to determine when I'm approaching the edges is when the behavior starts to diverge from my intuition.

Reply:
I agree. What I'm asking is: do you have any heuristics for knowing when using list monads/applicatives might behave in unintuitive ways?

Reply:
Personally, I wouldn't do anything more complex than several `<-` lines in a row, like:

```
foo xs = do
    x <- xs
    y <- getYs x
    z <- getZs x y
    pure $ bar x y z
```